### PR TITLE
[Master] Fix xml to record conversion for default value node

### LINKED ISSUE #42698: [Bug]: Xml to record conversion not working
### Description

Xml to record conversion is not working for below scenario,

```
<p:person xmlns:p="[www.example.org/personData](http://www.example.org/personData)" age="30">
    <n:name xmlns:n="[www.example.org/personName](http://www.example.org/personName)">John</n:name>
    <a:address xmlns:a="[www.example.org/personAddress](http://www.example.org/personAddress)">
        <c:city xmlns:c="[www.example.org/addressCity](http://www.example.org/addressCity)">Hollywood</c:city>
        <s:state xmlns:s="[www.example.org/addressState](http://www.example.org/addressState)">FL</s:state>
        <ct:country xmlns:ct="[www.example.org/addressCountry](http://www.example.org/addressCountry)">USA</ct:country>
    </a:address>
</p:person>
```

### Steps to Reproduce

_No response_

### Affected Version(s)

_No response_

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by mindula**
## Purpose
$subject.

Fixes #42698

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: misc/xml-to-record-converter/src/test/resources/ballerina/sample_28.bal [hunk 1]
```diff
@@ -0,0 +1,48 @@
+@xmldata:Namespace {prefix: "n", uri: "[www.example.org/personName](http://www.example.org/personName)"}
+type N_Name record {
+    string __text;
+    @xmldata:Attribute
+    string n = "[www.example.org/personName](http://www.example.org/personName)";
```
Comment (prakanth97) [inline misc/xml-to-record-converter/src/test/resources/ballerina/sample_28.bal:5]: Why we add this as attribute here?

File: misc/xml-to-record-converter/src/test/resources/ballerina/sample_28.bal [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+};
+
+@xmldata:Namespace {prefix: "c", uri: "[www.example.org/addressCity](http://www.example.org/addressCity)"}
+type C_City record {
+    string __text;
+    @xmldata:Attribute
+    string c = "[www.example.org/addressCity](http://www.example.org/addressCity)";
```
Comment (prakanth97) [inline misc/xml-to-record-converter/src/test/resources/ballerina/sample_28.bal:12]: We don't have to add the prefix to the fields according to the new implementation

Comment (mindula) [reply misc/xml-to-record-converter/src/test/resources/ballerina/sample_28.bal]: Fixed with [bfd1a23](https://github.com/ballerina-platform/ballerina-lang/pull/42699/commits/bfd1a239a8993c12bca58190302884efbea31f82)

**Review by LakshanWeerasinghe — APPROVED** [review]

File: misc/xml-to-record-converter/src/main/java/io/ballerina/xmltorecordconverter/XMLToRecordConverter.java [hunk 3]
```diff
@@ -321,6 +323,10 @@ private static List<Node> getRecordFieldsForXMLElement(Element xmlElement, boole
             }
         }
         if (isLeafXMLElementNode(xmlElement) && xmlElement.getAttributes().getLength() > 0) {
+            if (xmlElement.getAttributes().getLength() == 1 && xmlElement.getAttributes().item(0).getPrefix() != null &&
+                    XMLNS_PREFIX.equals(xmlElement.getAttributes().item(0).getPrefix())) {
```
Comment (prakanth97) [inline misc/xml-to-record-converter/src/main/java/io/ballerina/xmltorecordconverter/XMLToRecordConverter.java:327]:
```suggestion
            if (xmlElement.getAttributes().getLength() == 1 && xmlElement.getAttributes().item(0).getPrefix() != null 
                && XMLNS_PREFIX.equals(xmlElement.getAttributes().item(0).getPrefix())) {
```

File: misc/xml-to-record-converter/src/main/java/io/ballerina/xmltorecordconverter/XMLToRecordConverter.java [hunk 4]
```diff
@@ -248,7 +248,9 @@ private static List<Node> getRecordFieldsForXMLElement(Element xmlElement, boole
             if (xmlNode.getNodeType() == org.w3c.dom.Node.ELEMENT_NODE) {
                 Element xmlElementNode = (Element) xmlNode;
                 boolean isLeafXMLElementNode = isLeafXMLElementNode(xmlElementNode);
-                if (!isLeafXMLElementNode || xmlElementNode.getAttributes().getLength() > 0) {
+                if (!isLeafXMLElementNode || xmlElementNode.getAttributes().getLength() > 1 ||
+                        (xmlElementNode.getAttributes().getLength() == 1
+                                && !XMLNS_PREFIX.equals(xmlElementNode.getAttributes().item(0).getPrefix()))) {
```
Comment (prakanth97) [inline misc/xml-to-record-converter/src/main/java/io/ballerina/xmltorecordconverter/XMLToRecordConverter.java:253]:
```suggestion
                if (!isLeafXMLElementNode || xmlElementNode.getAttributes().getLength() > 1 
                    || (xmlElementNode.getAttributes().getLength() == 1
                                && !XMLNS_PREFIX.equals(xmlElementNode.getAttributes().item(0).getPrefix()))) {
```

File: misc/xml-to-record-converter/src/main/java/io/ballerina/xmltorecordconverter/XMLToRecordConverter.java [hunk 5]
```diff
@@ -330,8 +336,9 @@ private static List<Node> getRecordFieldsForXMLElement(Element xmlElement, boole
             recordFields.add(recordFieldNode);
             for (int j = 0; j < xmlElement.getAttributes().getLength(); j++) {
                 org.w3c.dom.Node xmlAttributeNode = xmlElement.getAttributes().item(j);
-                if (xmlAttributeNode.getNodeType() == org.w3c.dom.Node.ATTRIBUTE_NODE) {
-                    RecordFieldNode recordField = (RecordFieldNode) getRecordField(xmlAttributeNode);
+                if (xmlAttributeNode.getNodeType() == org.w3c.dom.Node.ATTRIBUTE_NODE && !XMLNS_PREFIX
+                        .equals(xmlAttributeNode.getPrefix())) {
```
Comment (prakanth97) [inline misc/xml-to-record-converter/src/main/java/io/ballerina/xmltorecordconverter/XMLToRecordConverter.java:340]:
```suggestion
                if (xmlAttributeNode.getNodeType() == org.w3c.dom.Node.ATTRIBUTE_NODE 
                    && !XMLNS_PREFIX.equals(xmlAttributeNode.getPrefix())) {
```

File: misc/xml-to-record-converter/src/main/java/io/ballerina/xmltorecordconverter/XMLToRecordConverter.java [hunk 6]
```diff
@@ -461,7 +468,9 @@ private static RecordFieldNode getRecordField(Element xmlElementNode, boolean is
         Token optionalFieldToken = isOptionalField ? questionMarkToken : null;
         Token semicolonToken = AbstractNodeFactory.createToken(SyntaxKind.SEMICOLON_TOKEN);
 
-        if (isLeafXMLElementNode(xmlElementNode) && xmlElementNode.getAttributes().getLength() == 0) {
+        if (isLeafXMLElementNode(xmlElementNode) && (xmlElementNode.getAttributes().getLength() == 0 ||
+                (xmlElementNode.getAttributes().getLength() == 1 &&
+                        XMLNS_PREFIX.equals(xmlElementNode.getAttributes().item(0).getPrefix())))) {
```
Comment (prakanth97) [inline misc/xml-to-record-converter/src/main/java/io/ballerina/xmltorecordconverter/XMLToRecordConverter.java:473]:
```suggestion
                (xmlElementNode.getAttributes().getLength() == 1 
                    && XMLNS_PREFIX.equals(xmlElementNode.getAttributes().item(0).getPrefix())))) {
```

File: misc/xml-to-record-converter/src/test/resources/xml/sample_33.xml [hunk 7]
```diff
@@ -0,0 +1,8 @@
+<p:person xmlns:p="[www.example.org/personData](http://www.example.org/personData)" age="30">
+    <n:name xmlns:n="[www.example.org/personName](http://www.example.org/personName)">John</n:name>
+    <a:address xmlns:a="[www.example.org/personAddress](http://www.example.org/personAddress)">
+        <c:city xmlns:c="[www.example.org/addressCity](http://www.example.org/addressCity)">Hollywood</c:city>
+        <s:state xmlns:s="[www.example.org/addressState](http://www.example.org/addressState)">FL</s:state>
+        <ct:country xmlns:ct="[www.example.org/addressCountry](http://www.example.org/addressCountry)">USA</ct:country>
+    </a:address>
+</p:person>
```
Comment (prakanth97) [inline misc/xml-to-record-converter/src/test/resources/xml/sample_33.xml:8]: Shall we add this as a test?
```
<p:person xmlns:p="[www.example.org/personData](http://www.example.org/personData)" age="30">
    <n:name xmlns:n="[www.example.org/personName](http://www.example.org/personName)" type="hello">John</n:name>
</p:person>
```

**Review by prakanth97 — DISMISSED** [review]
LGTM

File: misc/xml-to-record-converter/src/main/java/io/ballerina/xmltorecordconverter/XMLToRecordConverter.java [hunk 8]
```diff
@@ -248,7 +248,9 @@ private static List<Node> getRecordFieldsForXMLElement(Element xmlElement, boole
             if (xmlNode.getNodeType() == org.w3c.dom.Node.ELEMENT_NODE) {
                 Element xmlElementNode = (Element) xmlNode;
                 boolean isLeafXMLElementNode = isLeafXMLElementNode(xmlElementNode);
-                if (!isLeafXMLElementNode || xmlElementNode.getAttributes().getLength() > 0) {
+                if (!isLeafXMLElementNode || xmlElementNode.getAttributes().getLength() > 1
+                        || (xmlElementNode.getAttributes().getLength() == 1
+                                && !XMLNS_PREFIX.equals(xmlElementNode.getAttributes().item(0).getPrefix()))) {
```
Comment (KavinduZoysa) [inline misc/xml-to-record-converter/src/main/java/io/ballerina/xmltorecordconverter/XMLToRecordConverter.java:253]: Please extract this `xmlElementNode.getAttributes()` a variable, and check other places too.

**Review by prakanth97 — APPROVED** [review]

**Pull request merged**